File: ResolucaoProvas/2024.2/q3.py
from threading import Semaphore, Thread
from time import sleep
from random import shuffle
import uuid

SIZE = 1000
N_PRODUCERS = 4
N_CONSUMERS = 2

class Broker:
    requests = []
    producer = Semaphore(SIZE)
    consumer = Semaphore(0)
    
    def submitRequest(self, r):
        self.producer.acquire()
        self.requests.append(r)
        print(f"Master is appending {r} to requests") 
        self.consumer.release()

    def getWork(self):
        self.consumer.acquire()
        request = self.requests.pop()
        print(f"Worker is processing {request}")
        self.producer.release()

broker = Broker()

def workerHandler():
    global broker
    while (True):
        broker.getWork()
        sleep(1)

def masterHandler():
    global broker
    while (True):
        broker.submitRequest(uuid.uuid1())
        sleep(1)

if __name__ == "__main__":
    threads = []

    for i in range(N_PRODUCERS):
        threads.append(Thread(target=masterHandler, name=f"Master {i}"))
    
    for i in range(N_CONSUMERS):
        threads.append(Thread(target=workerHandler, name=f"Worker {i}"))
    

    shuffle(threads)

    for thread in threads:
        thread.start()
    
    for thread in threads:
        thread.join()

File: ResolucaoProvas/2024.2/Prova1/C/q3.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <semaphore.h>
#include <unistd.h>
#include <time.h>

#define SIZE 1000
#define N_PRODUCERS 4
#define N_CONSUMERS 2

typedef struct {
    int* requests;
    int count;
    sem_t producer;
    sem_t consumer;
} Broker;

Broker broker;

void broker_init() {
    broker.requests = (int*)malloc(SIZE * sizeof(int));
    broker.count = 0;
    sem_init(&broker.producer, 0, SIZE);
    sem_init(&broker.consumer, 0, 0);
}

void submitRequest(int r) {
    sem_wait(&broker.producer);
    broker.requests[broker.count] = r;
    broker.count++;
    printf("Master is appending %d to requests\n", r);
    sem_post(&broker.consumer);
}

void getWork() {
    sem_wait(&broker.consumer);
    broker.count--;
    int request = broker.requests[broker.count];
    printf("Worker is processing %d\n", request);
    sem_post(&broker.producer);
}

void* workerHandler(void* arg) {
    while(1) {
        getWork();
        sleep(1);
    }
    return NULL;
}

void* masterHandler(void* arg) {
    while(1) {
        submitRequest(rand());
        sleep(1);
    }
    return NULL;
}

int main() {
    srand(time(NULL));
    pthread_t threads[N_PRODUCERS + N_CONSUMERS];
    pthread_t temp_threads[N_PRODUCERS + N_CONSUMERS];

    broker_init();

    for(int i = 0; i < N_PRODUCERS; i++) {
        pthread_create(&temp_threads[i], NULL, masterHandler, NULL);
    }

    for(int i = 0; i < N_CONSUMERS; i++) {
        pthread_create(&temp_threads[N_PRODUCERS + i], NULL, workerHandler, NULL);
    }

    for(int i = 0; i < N_PRODUCERS + N_CONSUMERS; i++) {
        threads[i] = temp_threads[i];
    }

    for(int i = N_PRODUCERS + N_CONSUMERS - 1; i > 0; i--) {
        int j = rand() % (i + 1);
        pthread_t temp = threads[i];
        threads[i] = threads[j];
        threads[j] = temp;
    }

    for(int i = 0; i < N_PRODUCERS + N_CONSUMERS; i++) {
        pthread_join(threads[i], NULL);
    }

    free(broker.requests);
    sem_destroy(&broker.producer);
    sem_destroy(&broker.consumer);

    return 0;
}
